src/compiler/scala/tools/nsc/transform/async/AnfTransform.scala

```diff
@@ -136,19 +136,11 @@ private[async] trait AnfTransform extends TransformUtils {
           }
 
         case ValDef(mods, name, tpt, rhs) => atOwner(tree.symbol) {
-          // Capture current cursor of a non-empty `stats` buffer so we can efficiently restrict the
-          // `changeOwner` to the newly added items...
-          var statsIterator = if (currentStats.isEmpty) null else currentStats.iterator
-
           val expr = atOwner(currentOwner.owner)(transform(rhs))
 
-          // But, ListBuffer.empty.iterator doesn't reflect later mutation. Luckily we can just start
-          // from the beginning of the buffer
-          if (statsIterator == null) statsIterator = currentStats.iterator
-
```

Comment:
The intent of this code was to avoid `O(N^2)` cost for this transform wrt the number of statements in a block. The fail-fast iterators scupper the approach, but we can at least reduce the constant factor signficantly by avoiding the `changeOwner` for all but the new entries in the buffer:

```scala
        case ValDef(mods, name, tpt, rhs) => atOwner(tree.symbol) {
          val currentStatsLen = currentStats.length
          val expr = atOwner(currentOwner.owner)(transform(rhs))
          // Definitions within stats lifted out of the `ValDef` rhs should no longer be owned by the
          // the ValDef.
          currentStats.iterator.drop(currentStatsLen).foreach(_.changeOwner((currentOwner, currentOwner.owner)))
```

Reply:
in the previous code, the iterator was always created from the beginning `currentStats`, regardless of which expression created the iterator. since (as far as I can tell) we only ever append to `currentStats`, it still ended up iterating over all of it anyway. the comment

> Capture current cursor of a non-empty `stats` buffer so we can efficiently restrict the `changeOwner` to the newly added items...

was not, as far as I can tell, accurate in any way.

your change looks like it should work well though, and I will add it.